doc/src/build/programming-with-objects/ch6-collections.md

```diff
@@ -0,0 +1,189 @@
+    table: &Table<K, V>,
+    k: K
+): &V;
+
+public fun borrow_mut<K: copy + drop + store, V: store>(
+    table: &mut Table<K, V>,
+    k: K
+): &mut V;
+
+public fun remove<K: copy + drop + store, V: store>(
+    table: &mut Table<K, V>,
+    k: K,
+): V;
+
+}
+```
+
+These functions, add, read, write, and remove entries from the collection, respectively, and all accept keys by value. `Table` has type parameters for `K` and `V` so it is not possible to call these functions with different instantiations of `K` and `V` on the same instance of `Table`, however `Bag` does not these type parameters, and so does permit calls with different instantiations on the same instance.
+
+> :warning: Like with dynamic fields, it is an error to attempt to overwrite an existing key, or access or remove a non-existent key.
+
+> :warning: The extra flexibility of `Bag`'s heterogeneity means the type system will not statically prevent attempts to add a value with one type, and then borrow or remove it at another type. This pattern will fail at runtime with an abort, similar to the behavior for dynamic fields.
+
+### Querying Length
+
+It is possible to query all collection types for their length and check whether they are empty using the following family of functions:
+
+```rust
+module sui::table {
+
+public fun length<K: copy + drop + store, V: store>(
+    table: &Table<K, V>,
+): u64;
+
+public fun is_empty<K: copy + drop + store, V: store>(
+    table: &Table<K, V>
+): bool;
+
+}
+```
+
+> :bulb: `Bag` has these APIs, but they are not generic on `K` and `V` because `Bag` does not have these type parameters.
+
+### Querying for Containment
+
+All collections can be queried for key containment with:
+
+```rust
+module sui::table {
+
+public fun contains<K: copy + drop + store, V: store>(
+    table: &Table<K, V>
+    k: K
+): bool;
+
+}
+```
+
+The equivalent function for `Bag` is,
+
+```rust
+module sui::bag {
+
+public fun contains_with_type<K: copy + drop + store, V: store>(
+    bag: &Bag,
+    k: K
+): bool;
+
+}
+```
+
+which tests whether `bag` contains a key-value pair with key `k: K` and some value of type `V`.
+
+### Clean-up
+
+As mentioned in the introduction, collection types protect against accidental deletion when they might not be empty. This protection comes from the fact that they do not have `drop`, so must be explicitly deleted, using this API:
+
+```rust
+module sui::table {
+
+public fun destroy_empty<K: copy + drop + store, V: store>(
+    table: Table<K, V>,
+);
+
+}
+```
+
+This function takes the collection by value. If it contains no entries, it will be deleted, othewise the call will abort. `sui::table::Table` also has a convenience function,
+
+```
+module sui::table {
+
+public fun drop<K: copy + drop + store, V: drop + store>(
+    table: Table<K, V>,
+);
+
+}
+```
+
+that can only be called for tables where the both the key *and* value type have `drop` which allows it to delete tables whether they are empty or not.
```

Comment:
I find the wording here just a bit confusing, as the only new bit is the value part. The `K` type always needs `drop`